### PR TITLE
feature: refresh access token api

### PR DESCRIPTION
**Pull request by kuoche1712003**
## Why need this change? / Root cause: 
- When my access token expires, I need to renew it.
## Changes made:
- Add refresh access token api.
- OAuth2 scope: add offline_access to get a refresh token.
## Test Scope / Change impact:
- none
## Issue
- resolves https://github.com/Game-as-a-Service/Lobby-Platform-Service/issues/113

### CODE REVIEW

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        .withClientRegistrationId(CLIENT_REGISTRATION_ID)
+        .principal(
+            UsernamePasswordAuthenticationToken(
+                User(principalName, "", emptyList()), null, emptyList()
+            )
+        )
+        .attribute(HttpServletRequest::class.java.name, this.nativeRequest)
+        .attribute(HttpServletResponse::class.java.name, this.nativeResponse)
+        .build()
+}
```
Comment (frankvicky) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:83]: 建議顯式宣告回傳型別

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            ?.first()
+            ?.takeIf { it.description.startsWith("Jwt expired at") }
+            ?.let { true }
+            ?: false
+
+    private fun String.toPrincipalName() = try {
+        jwtDecoder.decode(this).subject
+    } catch (e: JwtValidationException) {
+        e.takeIf { it.onlyExpiredAt() }
+            .let { JWTParser.parse(this).jwtClaimsSet.subject }
```
Comment (frankvicky) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:62]: 這邊應該需要 `?. `

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    } catch (e: Exception) {
+        null
+    }
+
+    private fun String.canRefreshToken() =
+        authorizedClientService.loadAuthorizedClient<OAuth2AuthorizedClient>(CLIENT_REGISTRATION_ID, this)
+            ?.takeIf { it.refreshToken != null }
+            ?.let { true }
+            ?: false
```
Comment (frankvicky) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:71]: NIT：這邊我自己感覺有點難以閱讀，考慮使用 `when` 或其他方式改寫嗎？

**Review by frankvicky — CHANGES_REQUESTED** [review]

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        val authentication = JwtAuthenticationToken(jwtDecoder.decode(authorizedClient.accessToken.tokenValue))
+        authorizedClientService.saveAuthorizedClient(authorizedClient, authentication)
+    }
+
+    private fun JwtValidationException.onlyExpiredAt() =
+        errors.takeIf { it.size == 1 }
+            ?.first()
+            ?.takeIf { it.description.startsWith("Jwt expired at") }
+            ?.let { true }
+            ?: false
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:56]: `?.description?.startsWith("Jwt expired at") ?: false`

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+        return authorizedClient.accessToken.tokenValue
+    }
+
+    private fun saveAuthorizedClient(authorizedClient: OAuth2AuthorizedClient){
+        val authentication = JwtAuthenticationToken(jwtDecoder.decode(authorizedClient.accessToken.tokenValue))
+        authorizedClientService.saveAuthorizedClient(authorizedClient, authentication)
+    }
+
+    private fun JwtValidationException.onlyExpiredAt() =
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:51]: 放在調用行底下

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    private fun String.toPrincipalName() = try {
+        jwtDecoder.decode(this).subject
+    } catch (e: JwtValidationException) {
+        e.takeIf { it.onlyExpiredAt() }
+            .let { JWTParser.parse(this).jwtClaimsSet.subject }
+    } catch (e: Exception) {
+        null
+    }
+
+    private fun String.canRefreshToken() =
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:67]: ditto 放在調用行底下

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt (hunk 3 above)
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:71]: `?.refreshToken != null ?: false`

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    private fun NativeWebRequest.toOAuth2AuthorizeRequest(principalName: String) = OAuth2AuthorizeRequest
+        .withClientRegistrationId(CLIENT_REGISTRATION_ID)
+        .principal(
+            UsernamePasswordAuthenticationToken(
+                User(principalName, "", emptyList()), null, emptyList()
+            )
+        )
+        .attribute(HttpServletRequest::class.java.name, this.nativeRequest)
+        .attribute(HttpServletResponse::class.java.name, this.nativeResponse)
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:81]: 不需要用 this.

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/OAuth2Controller.kt [hunk 8]
```diff
@@ -22,6 +25,13 @@ class OAuth2Controller(
         return principal.tokenValue ?: "index"
     }
 
+    @PostMapping("/authenticate")
+    fun authenticate(request: NativeWebRequest, @RequestBody payload: AuthenticateToken): AuthenticateToken{
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/OAuth2Controller.kt:29]: 參數換行一下

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    private val authorizedClientService: OAuth2AuthorizedClientService,
+    private val manger: OAuth2AuthorizedClientManager,
+) {
+    companion object{
+        private const val CLIENT_REGISTRATION_ID = "auth0"
+    }
+
+    fun refreshAccessToken(
+        accessTokenValue: String,
+        request: NativeWebRequest
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:30]: 交換參數順序

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    } catch (e: Exception) {
+        null
+    }
+
+    private fun JwtValidationException.onlyExpiredAt(): Boolean =
+        errors.takeIf { it.size == 1 }
+            ?.first()
+            ?.description?.startsWith("Jwt expired at") ?: false
+
+
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/configs/securities/RefreshAccessTokenHandler.kt:69]: 空一行就好

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Review by frankvicky — APPROVED** [review]

**Pull request merged**